src/frontend/src/optimizer/plan_node/stream_group_topn.rs

```diff
@@ -51,8 +52,22 @@ impl StreamGroupTopN {
             watermark_columns
         };
 
-        let base = PlanBase::new_stream_with_logical(
-            &logical,
+        let mut stream_key = logical
+            .stream_key()
+            .expect("logical node should have stream key here");
+        if let Some(vnode_col_idx) = vnode_col_idx && stream_key.len() > 1 {
+            // The output stream key of `GroupTopN` is a union of group key and input stream key,
+            // while vnode is calculated from a subset of input stream key. So we can safely remove
+            // the vnode column from output stream key. While at meanwhile we cannot leave the stream key
+            // as empty, so we only remove it when stream key length is > 1.
+            stream_key.remove(stream_key.iter().position(|i| *i == vnode_col_idx).unwrap());
```

Comment:
I'm trying to see the changes caused by this line, but found after commenting out this line, it panics with "a stream key is expected but not exist". Why it's the case? 🤔

Reply:
Yes, this line is to trim the `vnode` column added just for 2-phase optimization. Because it's added to `group by`, it must appear in `stream key` in the middle phase, while at the same time we need a `Project` at the end of this subplan to remove `vnode`. The problem here has two aspects:

1. `Project` cannot trim unneeded part of stream key. All columns in stream key should appear in `exprs`;
2. `vnode` is not necessarily a semantical part of stream key, we added it just because it's in `group by` of the middle phase `GroupTopN`.

So the solution here is to fix the 2nd aspect - just remove it from stream key when possible. For the cases where original stream key is empty, it must have `limit` equals to 1, then the final phase won't inherit `vnode` as part of it's stream key, and hence `Project` won't see it.

Reply:
Nice explanation! But I'm still confused why it didn't break before this PR?

Reply:
Ohhhh so here's the hack
<img width="783" alt="image" src="https://github.com/risingwavelabs/risingwave/assets/37948597/bbaa6be3-aa7a-4f9b-a030-fa40ac362d67">

Reply:
> Ohhhh so here's the hack
> <img alt="image" width="783" src="https://user-images.githubusercontent.com/37948597/275421572-bbaa6be3-aa7a-4f9b-a030-fa40ac362d67.png">

Exactly, that's the previous hack.

Reply:
Short explanation:

If we don't remove 
```
Global TopN
|
Project        stream key: None (because part of stream key (vnode) is absent)
|
Local TopN     stream key: [..., vnode]
```